### ISSUE TITLE
[Bug] Prettier v3 Support

### ISSUE THREAD

**Issue by evoactivity** [gitKrystan/prettier-plugin-ember-template-tag#61]
### 🐞 Describe the Bug

Trying to get this working in vs code, also seeing this error when running `pnpm lint`. 

Sidenote, to get to this point I had to follow the tips in https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/38

### 🔬 Minimal Reproduction
Not sure how reproducible this is I'm not sure of the cause. As it shows up outside of vscode I don't think it's an editor specific issue.

Some probably relevant info

- Working in a pnpm monorepo setup by turbo.
- Fresh ember project (5.1.2)

### 😕 Actual Behavior

Will not run. Failing with `assertExists failed` and a stack trace of minified code.

### 🤔 Expected Behavior

Prettier is run.

### 🌍 Environment

- prettier-plugin-ember-template-tag version: 0.3.2
- ember-template-imports version: 3.4.2
- prettier version: 3.0.0

### ➕ Additional Context
Output from VS Code
```
["INFO" - 12:52:44 AM] Formatting file:///Users/liam/Work/project/monorepo/apps/web/app/components/test-thing.gts
["INFO" - 12:52:44 AM] Using config file at '/Users/liam/Work/project/monorepo/apps/web/.prettierrc.js'
["INFO" - 12:52:44 AM] PrettierInstance:
{
  "modulePath": "/Users/liam/Work/project/monorepo/apps/web/node_modules/prettier/index.cjs",
  "importResolver": {},
  "callMethodResolvers": {},
  "currentCallMethodId": 2,
  "version": "3.0.0"
}
["INFO" - 12:52:44 AM] Using ignore file (if present) at /Users/liam/Work/project/monorepo/.prettierignore
["INFO" - 12:52:44 AM] File Info:
{
  "ignored": false,
  "inferredParser": "ember-template-tag"
}
["INFO" - 12:52:44 AM] Detected local configuration (i.e. .prettierrc or .editorconfig), VS Code configuration will not be used
["INFO" - 12:52:44 AM] Prettier Options:
{
  "filepath": "/Users/liam/Work/project/monorepo/apps/web/app/components/test-thing.gts",
  "parser": "ember-template-tag",
  "useTabs": false,
  "tabWidth": 2,
  "endOfLine": "lf",
  "semi": true,
  "singleQuote": true,
  "quoteProps": "consistent",
  "bracketSpacing": true,
  "arrowParens": "always",
  "printWidth": 120,
  "plugins": [
    "/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js"
  ],
  "importOrder": [
    "web",
    "^node:(.*)$",
    "^@glimmer/(.*)$",
    "^@ember/(.*)$",
    "<THIRD_PARTY_MODULES>",
    "^[./]"
  ],
  "importOrderSeparation": false,
  "importOrderSortSpecifiers": true,
  "importOrderGroupNamespaceSpecifiers": true,
  "importOrderCaseInsensitive": true,
  "importOrderParserPlugins": [
    "typescript",
    "[\"decorators\", { \"decoratorsBeforeExport\": true }]"
  ]
}
["ERROR" - 12:52:44 AM] Error formatting document.
["ERROR" - 12:52:44 AM] assertExists failed
Error: assertExists failed
    at Bp (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:1933:57639)
    at TA (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:1933:57698)
    at FM (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:2048:8267)
    at Object.preprocess (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:2048:11036)
    at parse (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:17076:43)
    at async coreFormat (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:18401:7)
    at async formatWithCursor (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:18603:14)
    at async Module.format2 (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:21789:25)
["INFO" - 12:52:44 AM] Formatting completed in 13ms.
```

Output from `pnpm lint:js`

```
$ pnpm lint:js           

> project-web@0.0.0 lint:js /Users/liam/Work/project/monorepo/apps/web
> eslint . --cache


Oops! Something went wrong! :(

ESLint: 8.44.0

Error: assertExists failed
    at Bp (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:1933:57639)
    at TA (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:1933:57698)
    at FM (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:2048:8267)
    at Object.preprocess (/Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier-plugin-ember-template-tag@0.3.2/node_modules/prettier-plugin-ember-template-tag/dist/prettier-plugin-ember-template-tag.js:2048:11036)
    at parse (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:17076:43)
    at async coreFormat (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:18401:7)
    at async formatWithCursor (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:18603:14)
    at async Module.format2 (file:///Users/liam/Work/project/monorepo/node_modules/.pnpm/prettier@3.0.0/node_modules/prettier/index.mjs:21789:25)
```

**Comment by evoactivity** [comment]
Downgrading prettier to 2.8.7 and eslint-plugin-prettier to 4.2.1 got it running. So seems to be an issue with pretter 3

**Comment by ef4** [comment]
I looked into this a little bit. It's non-trivial. 

The proximate cause of the failure is that `assertExists( estreePrinter.hasPrettierIgnore )` fails because `estreePrinter` is now an async function that returns the printer, not the printer itself. There's not an obvious place to absorb the asynchrony right at that spot. 

The call signature of `embed` has also changed so that:

```js
printer.embed = (path, print, textToDoc, options) => {
  return stuff
}
```

becomes something like:

```js
printer.embed = (path, options) => {
  return async (textToDoc, print) => {
    return stuff;
  }
}
```

```

**Comment by Techn1x** [comment]
Hitting the same thing error on updating to prettier v3

> There's not an obvious place to absorb the asynchrony right at that spot.

If we go up a few calls, the code is consumed in a `preprocess` hook - prettier v3 says 
"Also, the preprocess method of a printer can return a promise now."

So I think that'll be the place.

**Comment by Techn1x** [comment]
I've opened a draft PR here
https://github.com/gitKrystan/prettier-plugin-ember-template-tag/pull/63

I am stuck on 2 types issues, if anyone could lend a hand. Else, feel free to use it as a guide for your own prettier v3 update

**Comment by gitKrystan** [comment]
I've been watching the status updates for Prettier v3 like 🙈 . Looks like it's here now. I'll try to look into this ASAP

**Comment by esbanarango** [comment]
Facing this same issue trying out `ember-template-imports`. This project already uses `prettierV3`.

<img width="1499" alt="index_gts_—_Untitled__Workspace_" src="https://github.com/gitKrystan/prettier-plugin-ember-template-tag/assets/1156865/0c7a901b-a1a0-48da-86bf-ea6d54f0df63">

**Comment by gitKrystan** [comment]
Just released `prettier-plugin-ember-template-tag@1.0.0-0` with a fix for this issue. Can one of you try it out and LMK if everything is working properly with your set-up? After that I will release `1.0.0` for real.

**Comment by HeroicEric** [comment]
@gitKrystan that is working for me with prettier 3.0.0 🚀 🎉 🚀 🎉 🚀 🎉

**Comment by esbanarango** [comment]
Same here!!! That fix solved the issue.

(Now I'm blocked by https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/81#issuecomment-1650560485 😂 )

**Comment by evoactivity** [comment]
Working for me. @gitKrystan Thanks for fixing this so quickly 😊

**Comment by gitKrystan** [comment]
Fixed as of v1.0.0 🎉

**Comment by Techn1x** [comment]
works great for me as well. thanks for putting in the effort here! happily on prettier 3 now :)
